src/sketch.lisp

```diff
@@ -182,136 +182,126 @@ used for drawing, 60fps.")
+
+(defun binding= (a b)
+  (equalp (slot-value a 'name) (slot-value b 'name)))
+
+(defun make-binding (sketch-name list)
+  (let* ((name (car list))
+	 (channelp (and (consp (cadr list)) (eql 'in (caadr list))))
+	 (original-value (cadr list))
+	 (value (if channelp (third (cadr list)) (cadr list))) ; lifts default channel value
+	 (defaultp (and (member name (mapcar #'car *default-slots*) :test #'equalp) t))
+	 (initarg (alexandria:make-keyword name))
+	 (accessor (if defaultp
+		       (intern (string (alexandria:symbolicate 'sketch- name)) :sketch)
+		       (alexandria:symbolicate sketch-name '- name))))
+    (make-instance 'binding :sketch-name sketch-name
+			    :name name
+			    :channelp channelp
+			    :original-value original-value
+			    :value value
+			    :defaultp defaultp
+			    :initarg initarg
+			    :accessor accessor)))
+
+(defun make-binding-from-default-slot (slot)
+  (let ((name (car slot))
+	(value (cadr (member :initform slot)))
+	(defaultp t)
+	(accessor (cadr (member :accessor slot)))
+	(initarg (cadr (member :initarg slot))))
+    (make-instance 'binding :sketch-name 'sketch
+			    :name name
+			    :channelp nil
+			    :original-value value
+			    :value value
+			    :defaultp defaultp
+			    :initarg initarg
+			    :accessor accessor)))
+
+(defmethod binding->slot ((binding binding))
+  `(,(binding-name binding)
+    :initarg ,(alexandria:make-keyword (binding-name binding))
+    :accessor ,(binding-accessor binding)))
+
+ ;;; DEFSKETCH macro
+
+(defun define-sketch-defclass (sketch-name custom-bindings)
+  `(defclass ,sketch-name (sketch)
+     ,(mapcar #'binding->slot custom-bindings)))
+
+
+(defun define-channel-observers (all-bindings)
+  (let ((channel-bindings (remove-if-not #'binding-channel-p all-bindings)))
+    (loop for binding in channel-bindings
+	  collect
+	  `(define-channel-observer
+	     (let ((win (kit.sdl2:last-window)))
+	       (when win
+		 (setf (,(binding-accessor binding) win) ,(binding-original-value binding))))))))
+
+  (defun define-draw-method (sketch-name all-bindings body)
+    `(defmethod draw ((*sketch* ,sketch-name) &key &allow-other-keys)
+       (with-accessors ,(mapcar (lambda (binding)
+				  (list (binding-name binding) (binding-accessor binding)))
+			 all-bindings)
+	   *sketch*
+	 ,@body)))
+
+(defun define-prepare-method (sketch-name default-bindings defined-bindings custom-bindings all-bindings)
+  (flet ((binding-in (bindings)
+	   (lambda (binding) (member binding bindings :test #'binding=))))
+    (let ((changed-bindings (remove-if-not (binding-in default-bindings) defined-bindings))
+	  (unchanged-bindings (remove-if (binding-in defined-bindings) default-bindings)))
+      `(defmethod prepare ((*sketch* ,sketch-name) &rest initargs &key &allow-other-keys)
+	 (declare (ignorable initargs))
+	 (let* (,@(loop for b in unchanged-bindings
+			collect `(,(binding-name b) (slot-value *sketch* ',(binding-name b))))
```

Comment:
For `unchanged-bindings` you could use the same "formula" as for `changed-bindings` with the default value `(binding-value b)` being a literal - string, symbol or a number - so it won't be executed twice for no reason.